File: 2022-2023/sourcecode/lab/lab11/RetrofitUrlBasicsJava/app/src/main/java/pl/udu/uwr/pum/retrofiturlbasicsjava/RetrofitFactory.java
package pl.udu.uwr.pum.retrofiturlbasicsjava;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public final class RetrofitFactory {
    private RetrofitFactory(){}

    private static volatile PlaceholderService INSTANCE;
    private final String url = "https://jsonplaceholder.typicode.com/";

    public static PlaceholderService makeService() {
        if (INSTANCE == null) {
            synchronized (RetrofitFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new Retrofit.Builder()
                            .baseUrl("https://jsonplaceholder.typicode.com/")
                            .addConverterFactory(GsonConverterFactory.create())
                            .build().create(PlaceholderService.class);
                }
            }
        }
        return INSTANCE;
    }
}

File: 2022-2023/sourcecode/lab/lab11/RetrofitPutPatchDeleteBasicsKotlin/app/src/main/java/pl/udu/uwr/pum/retrofitputpatchdeletebasicskotlin/RetrofitFactory.kt
package pl.udu.uwr.pum.retrofiturlbasicskotlin

import okhttp3.OkHttpClient
import pl.udu.uwr.pum.retrofitputpatchdeletebasicskotlin.PlaceholderService
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitFactory {

    private const val url = "https://jsonplaceholder.typicode.com/"

    val service: PlaceholderService by lazy {
        Retrofit.Builder()
            .baseUrl(url)
            .addConverterFactory(GsonConverterFactory.create())
            .build().create(PlaceholderService::class.java)
    }
}